dali/pipeline/operator/sequence_operator.h

```diff
@@ -0,0 +1,516 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_OPERATOR_SEQUENCE_OPERATOR_H_
+#define DALI_PIPELINE_OPERATOR_SEQUENCE_OPERATOR_H_
+
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include "dali/pipeline/data/tensor.h"
+#include "dali/pipeline/data/views.h"
+#include "dali/pipeline/operator/argument.h"
+#include "dali/pipeline/operator/common.h"
+#include "dali/pipeline/operator/op_spec.h"
+#include "dali/pipeline/operator/operator.h"
+#include "dali/pipeline/operator/sequence_shape.h"
+
+namespace dali {
+
+template <typename Backend>
+class SequenceOperator : public Operator<Backend> {
+ public:
+  inline explicit SequenceOperator(const OpSpec &spec) : Operator<Backend>{spec} {}
+
+  using Operator<Backend>::Setup;
+  using Operator<Backend>::Run;
+
+  template <typename InputBackend>
+  using ws_input_t = typename workspace_t<Backend>::template input_t<InputBackend>;
+  template <typename OutputBackend>
+  using ws_output_t = typename workspace_t<Backend>::template output_t<OutputBackend>;
+
+  bool Setup(std::vector<OutputDesc> &output_desc, const workspace_t<Backend> &ws) override {
+    SetupSequenceOperator(ws);
+    is_expanding_ = ShouldExpand(ws);
+    bool is_inferred;
+    if (!IsExpanding()) {
+      is_inferred = Operator<Backend>::Setup(output_desc, ws);
+    } else {
+      DALI_ENFORCE(ExpandLikeIdx() >= 0 &&
+                   GetInputExpandDesc(ExpandLikeIdx()).NumDimsToExpand() > 0);
+      SetupExpandedWorkspace(expanded_, ws);
+      ExpandInputs(ws);
+      ExpandArguments(ws);
+      is_inferred = Operator<Backend>::Setup(output_desc, expanded_);
```

Comment:
I might be highly nitpicking but maybe this should be something like:
```suggestion
      const auto &expanded = SetupExpandedWorkspace(ws);  // include ExapndInputs and ExpandArguments here?
      is_inferred = Operator<Backend>::Setup(output_desc, expanded);
```

Reply:
Don't know how it would fit into Run though? Similar accessor returning const ref for the purpose of Run?

Reply:
Yeah, definitely we need to keep the workspace alive between Setup and Run somehow, so keeping it as a member seems like a good option. However, putting the calls to ExpandInputs and ExpandArguments inside SetupExpandedWorkspace has some merit.

Reply:
But in the `Run` method we would still have ExpandOutputs from the expand-stuff-in-the-workspace family. To me keeping it top-level here adds some readability to the operator idea.